### PR TITLE
Add Parser.choice

### PR DESCRIPTION
**Pull request by adithyaov**

### CODE REVIEW

File: test/Streamly/Test/Data/Parser.hs [hunk 1]
```diff
@@ -372,6 +372,25 @@ takeWhile1 =
...
+                      checkListEqual parsed_list $ take (min i j) ls
+                  Left _ -> property False
+
+    where
+
+    parser i = P.fromFold (FL.take i FL.toList)
+
+
+
+
```
Comment (harendra-kumar) [inline test/Streamly/Test/Data/Parser.hs:393]: Remove extra blank lines.

File: benchmark/Streamly/Benchmark/Data/Parser/ParserD.hs [hunk 2]
```diff
@@ -64,6 +64,20 @@ benchIOSink value name f =
...
+
+{-# INLINE choice2 #-}
+choice2 :: MonadCatch m => Int -> SerialT m a -> m ()
+choice2 value =
+    IP.parseD
+        (PR.choice
+             [ PR.fromFold (FL.take (value `div` 2) FL.drain)
+             , PR.fromFold (FL.take (value `div` 2) FL.drain)
+             ])
+
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Parser/ParserD.hs:80]: We already have a benchmark for `choice` which uses `asum`, we can rename that to `choice (asum)`. For this benchmark we should use exactly the same test as for `asum` to make a fair comparison. Also, just one benchmark is enough.

File: src/Streamly/Internal/Data/Parser.hs [hunk 3]
```diff
@@ -929,13 +929,11 @@ concatMap f p = K.toParserK $ D.concatMap (K.fromParserK . f) (K.fromParserK p)
...
 --
--- /Unimplemented/
+-- /Internal/
 --
 {-# INLINE choice #-}
-choice ::
-    -- Foldable t =>
-    t (Parser m a b) -> Parser m a b
-choice _ps = undefined
+choice :: (Functor t, Foldable t, MonadCatch m) => t (Parser m a b) -> Parser m a b
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Parser.hs:935]: Going beyond 80 cols.

File: src/Streamly/Internal/Data/Parser.hs [hunk 4]
```diff
@@ -929,13 +929,11 @@ concatMap f p = K.toParserK $ D.concatMap (K.fromParserK . f) (K.fromParserK p)
 --
 -- This is same as 'asum' but more efficient.
 --
--- /Unimplemented/
+-- /Internal/
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Parser.hs:932]: `Pre-release`.

File: test/Streamly/Test/Data/Parser.hs [hunk 5]
```diff
@@ -372,6 +372,25 @@ takeWhile1 =
         where
             predicate = (== 0)
 
+choice :: Property
+choice =
```
Comment (harendra-kumar) [inline test/Streamly/Test/Data/Parser.hs:376]: Can we test `asum` as well using this same test?

**Review by harendra-kumar — CHANGES_REQUESTED** [review]

File: src/Streamly/Internal/Data/Parser/ParserD.hs [hunk 6]
```diff
@@ -883,16 +883,14 @@ sequence _f _p = undefined
...
--- /Unimplemented/
+-- /Internal/
 --
 {-# INLINE choice #-}
-choice ::
-    -- Foldable t =>
-    t (Parser m a b) -> Parser m a b
-choice _ps = undefined
+choice :: (MonadCatch m, Foldable t) => t (Parser m a b) -> Parser m a b
+choice = foldl1 shortest
```
Comment (harendra-kumar) [inline src/Streamly/Internal/Data/Parser/ParserD.hs:893]: `shortest` is marked as `Broken`, so we need to mark this also broken unless we review and fix that.

File: src/Streamly/Internal/Data/Parser.hs (hunk 4 above)
Comment (adithyaov) [reply src/Streamly/Internal/Data/Parser.hs:932]: `Broken`

File: test/Streamly/Test/Data/Parser.hs (hunk 5 above)
Comment (adithyaov) [reply test/Streamly/Test/Data/Parser.hs:376]: Using the same test failed, so probably not. I have not looked into why it failed.

File: benchmark/Streamly/Benchmark/Data/Parser/ParserD.hs [hunk 7]
```diff
@@ -351,7 +357,8 @@ o_n_space_serial value =
     , benchIOSink value "sequenceA_/100" $ sequenceA_ (value `div` 100)
     , benchIOSink value "sequence/100" $ sequence (value `div` 100)
     , benchIOSink value "sequence_/100" $ sequence_ (value `div` 100)
-    , benchIOSink value "choice/100" $ choice (value `div` 100)
+    , benchIOSink value "choice (asum)/100" $ choiceAsum (value `div` 100)
+    , benchIOSink value "choice" $ choice (value `div` 100)
```
Comment (harendra-kumar) [inline benchmark/Streamly/Benchmark/Data/Parser/ParserD.hs:361]: Should name it to `choice/100` to indicate that this benchmark is 100 times shorter than usual.

File: test/Streamly/Test/Data/Parser.hs (hunk 5 above)
Comment (harendra-kumar) [reply test/Streamly/Test/Data/Parser.hs:376]: That means there is something wrong with this or with asum. We should file an issue if we are not investigating it right now.

**Review by harendra-kumar — APPROVED** [review]
There are a few minor comments.

**Review by harendra-kumar — APPROVED** [review]

File: test/Streamly/Test/Data/Parser.hs (hunk 5 above)
Comment (adithyaov) [reply test/Streamly/Test/Data/Parser.hs:376]: See https://github.com/composewell/streamly/issues/1191

**Pull request merged**